arch/risc-v/src/mpfs/mpfs_coremmc.c

```diff
@@ -0,0 +1,2410 @@
+
+static sdio_capset_t mpfs_capabilities(struct sdio_dev_s *dev)
+{
+  struct mpfs_dev_s *priv = (struct mpfs_dev_s *)dev;
+  sdio_capset_t caps = 0;
+
+  if (priv->onebit)
+    {
+      caps |= SDIO_CAPS_1BIT_ONLY;
+    }
+
+  return caps;
+}
+
+/****************************************************************************
+ * Name: mpfs_status
+ *
+ * Description:
+ *   Get SDIO status.
+ *
+ * Input Parameters:
+ *   dev   - Device-specific state data
+ *
+ * Returned Value:
+ *   Returns a bitset of status values (see mpfs_status_* defines)
+ *
+ ****************************************************************************/
+
+static sdio_statset_t mpfs_status(struct sdio_dev_s *dev)
+{
+  struct mpfs_dev_s *priv = (struct mpfs_dev_s *)dev;
+  return priv->cdstatus;
+}
+
+/****************************************************************************
+ * Name: mpfs_widebus
+ *
+ * Description:
+ *   Called after change in Bus width has been selected (via ACMD6).  Most
+ *   controllers will need to perform some special operations to work
+ *   correctly in the new bus mode.
+ *
+ * Input Parameters:
+ *   dev  - An instance of the SDIO device interface
+ *   wide - true: wide bus (4-bit) bus mode enabled
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void mpfs_widebus(struct sdio_dev_s *dev, bool wide)
+{
+  struct mpfs_dev_s *priv = (struct mpfs_dev_s *)dev;
+  priv->widebus = wide;
+
+  mcinfo("wide: %d\n", wide);
+}
+
+/****************************************************************************
+ * Name: mpfs_set_hs_mode
+ *
+ * Description:
+ *   Sets the device in HS mode via CMD6.
+ *
+ * Input Parameters:
+ *   dev    - An instance of the SDIO device interface
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void mpfs_set_hs_mode(struct sdio_dev_s *dev)
+{
+  uint32_t r1;
+  int ret;
+
+  if ((ret = mpfs_sendcmd(dev, MMCSD_CMD6, 0x03b90100u)) == OK)
+    {
+      if ((ret == mpfs_waitresponse(dev, MMCSD_CMD6)) == OK)
+        {
+          ret = mpfs_recvshortcrc(dev, MMCSD_CMD6, &r1);
+        }
+    }
+
+  if (ret < 0)
+    {
+      mcerr("Failed to set high speed mode\n");
+      return;
+    }
+}
+
+/****************************************************************************
+ * Name: mpfs_set_4bit_mode
+ *
+ * Description:
+ *   Sets the device for 4-bit data width via CMD6.
+ *
+ * Input Parameters:
+ *   dev    - An instance of the SDIO device interface
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void mpfs_set_4bit_mode(struct sdio_dev_s *dev)
+{
+  struct mpfs_dev_s *priv = (struct mpfs_dev_s *)dev;
+  uint32_t r1;
+  uint8_t dw;
+  int ret;
+
+  /* Need to settle a bit before issuing this CMD6 */
+
+  nxsig_usleep(100);
```

Comment:
Hmm, why is this delay needed ?

Reply:
Not sure, but it didn't work without. Could test with the latest, removing it out

Reply:
mpfs_waitresponse: ERROR: Timeout cmd: 00000086 stat: 44
mpfs_check_recverror: ERROR: Command timeout: 44
mpfs_set_4bit_mode: Failed to set 4-bit mode

I don't know honestly, just tried removing it out. Get errors like above.